File: 874-BackspaceStringCompare/874-BackspaceStringCompare.py
# Last updated: 8/16/2026, 9:49:16 PM
class Solution:
    def backspaceCompare(self, s: str, t: str) -> bool:
        # actually its STACK, FILO
        new_s_list, new_t_list = [], []

        for char in s:
            if char == '#': 
                new_s_list = new_s_list[:-1]
            else:
                new_s_list.append(char)
        
        new_s = ''.join(new_s_list)

        for char in t:
            if char == '#': 
                new_t_list = new_t_list[:-1]
                # do not use pop(), as it will cause pop in empty list
            else:
                new_t_list.append(char)
        
        new_t = ''.join(new_t_list)

        return new_s == new_t
        
        

File: 874-BackspaceStringCompare/874-BackspaceStringCompare.java
// Last updated: 8/16/2026, 9:49:20 PM
class Solution {
    public boolean backspaceCompare(String s, String t) {
        // FILO, Stack
        Stack<Character> s1 = new Stack<>(), s2 = new Stack<>();
        for(char c: s.toCharArray()) {
            if(c == '#' && !s1.isEmpty()) {
                s1.pop();
            } else if(c == '#' && s1.isEmpty()) {
                continue;
            } else {
                s1.push(c);
            }
        }
        for(char c: t.toCharArray()) {
            if(c == '#'&& !s2.isEmpty()) {
                s2.pop();
            } else if(c == '#' && s2.isEmpty()) {
                continue;
            } else {
                s2.push(c);
            }
        }
        return s1.equals(s2);
    }
}